### PR TITLE
Chan fix duplicate

### PR DESCRIPTION
**Pull request by angelwen08**

### CODE REVIEW

File: src/jsonTraverse.ts [hunk 1]
```diff
@@ -63,6 +63,18 @@ export function traverse(node: Node, parentUid: string, parentContainer: YamlMod
...
+        // to add this to handle duplicate class and module under the same hirachy
+        if (node.kindString === 'Class' || node.kindString === 'Interface') {
+            if (parentType === 'Class' || parentType === 'Interface') {
+                let currentUID = uid + `.${node.name}`;
+                let mapping: string[] = [];
+                if (innerClassReferenceMap.has(uid)) {
+                    mapping = innerClassReferenceMap.get(uid);               
+                } 
+                mapping.push(currentUID);
+                innerClassReferenceMap.set(uid, mapping);
```
Comment (wuhanyumsft) [inline src/jsonTraverse.ts]: we don't need to set the mapping if it's already in ReferenceMap

File: src/main.ts [hunk 2]
```diff
@@ -107,10 +108,22 @@ if (rootElements && rootElements.length) {
...
+            let referencedClass = innerClassReferenceMapping.get(transfomredClass.items[0].uid);
+            referencedClass.forEach(function (item) {
+              let names = item.split('.');
+              let ref: Reference = {
+                uid: item,
+                name: names[names.length - 1]
+              };
+              reference.push(ref);
+            });
+          }
```
Comment (wuhanyumsft) [inline src/main.ts]: indent

File: src/jsonTraverse.ts [hunk 3]
```diff
@@ -63,6 +63,18 @@ export function traverse(node: Node, parentUid: string, parentContainer: YamlMod
     }
 
     if ((node.kindString === 'Class' || node.kindString === 'Interface' || node.kindString === 'Enumeration' || node.kindString === 'Type alias') && node.name) {
+        // to add this to handle duplicate class and module under the same hirachy
+        if (node.kindString === 'Class' || node.kindString === 'Interface') {
+            if (parentType === 'Class' || parentType === 'Interface') {
+                let currentUID = uid + `.${node.name}`;
+                let mapping: string[] = [];
```
Comment (wuhanyumsft) [inline src/jsonTraverse.ts]: It's not a mapping. So other name may be more accurate

File: src/jsonTraverse.ts [hunk 4]
```diff
@@ -63,6 +63,18 @@ export function traverse(node: Node, parentUid: string, parentContainer: YamlMod
     }
 
     if ((node.kindString === 'Class' || node.kindString === 'Interface' || node.kindString === 'Enumeration' || node.kindString === 'Type alias') && node.name) {
+        // to add this to handle duplicate class and module under the same hirachy
+        if (node.kindString === 'Class' || node.kindString === 'Interface') {
+            if (parentType === 'Class' || parentType === 'Interface') {
+                let currentUID = uid + `.${node.name}`;
```
Comment (wuhanyumsft) [inline src/jsonTraverse.ts]: `uid.${node.name}`

File: src/main.ts [hunk 5]
```diff
@@ -107,10 +108,22 @@ if (rootElements && rootElements.length) {
     }, []);
 
     insertClassReferenceForModule(flattenElements);
-
     console.log('Yaml dump start.');
     fs.ensureDirSync(outputPath);
     flattenElements.forEach(transfomredClass => {
+        // to add this to handle duplicate class and module under the same hirachy
+        if (transfomredClass.items[0].type === 'class' && innerClassReferenceMapping.has(transfomredClass.items[0].uid)) {
```
Comment (wuhanyumsft) [inline src/main.ts]: class or interface?

File: src/main.ts [hunk 6]
```diff
@@ -107,10 +108,22 @@ if (rootElements && rootElements.length) {
...
+            let reference = transfomredClass.references;
+            let referencedClass = innerClassReferenceMapping.get(transfomredClass.items[0].uid);
```
Comment (wuhanyumsft) [inline src/main.ts]: referencingClasses

File: src/main.ts [hunk 7]
```diff
@@ -107,10 +108,22 @@ if (rootElements && rootElements.length) {
...
+            referencedClass.forEach(function (item) {
+              let names = item.split('.');
+              let ref: Reference = {
+                uid: item,
```
Comment (wuhanyumsft) [inline src/main.ts]: change the hack the logic into a method for better structure.

File: src/jsonTraverse.ts [hunk 8]
```diff
@@ -63,6 +63,18 @@ export function traverse(node: Node, parentUid: string, parentContainer: YamlMod
     }
 
     if ((node.kindString === 'Class' || node.kindString === 'Interface' || node.kindString === 'Enumeration' || node.kindString === 'Type alias') && node.name) {
+        // to add this to handle duplicate class and module under the same hirachy
```
Comment (zhiliangxu) [inline src/jsonTraverse.ts]: typo: hirachy

**Pull request merged**